### PR TITLE
system-program: Remove zero lamport check on transfers

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

It's possible to bypass signature checks if transferring 0 lamports, causing potential unclear / misleading behavior.

#### Summary of Changes

Perform all checks on transfers, even for 0 lamports.

Fixes #

### CODE REVIEW

File: runtime/src/system_instruction_processor.rs [hunk 1]
```diff
@@ -179,6 +179,13 @@ fn transfer_verified(
         ic_msg!(invoke_context, "Transfer: `from` must not carry data");
         return Err(InstructionError::InvalidArgument);
     }
+
+    if invoke_context.is_feature_active(&feature_set::system_transfer_zero_check::id())
+        && lamports == 0
+    {
+        return Ok(());
+    }
+
```
Comment (jstarry) [inline runtime/src/system_instruction_processor.rs:188]: this looks unnecessary. The lamports check below will be bypassed when `lamports == 0`

File: runtime/src/system_instruction_processor.rs [hunk 2]
```diff
@@ -666,21 +677,21 @@ mod tests {
...
-        let to = solana_sdk::pubkey::new_rand();
+        let to = Pubkey::new_unique();
         let to_account = AccountSharedData::new_ref(0, 0, &Pubkey::default());
 
         assert_eq!(
             create_account(
-                &KeyedAccount::new(&from, false, &from_account), // no signer
-                &KeyedAccount::new(&to, false, &to_account),
+                &KeyedAccount::new(&from, true, &from_account),
+                &KeyedAccount::new(&to, true, &to_account),
```
Comment (jstarry) [inline runtime/src/system_instruction_processor.rs:689]: doesn't need to be a signer

Comment (jstarry) [reply runtime/src/system_instruction_processor.rs:689]: oops, I thought this was just a transfer, never mind

**Review by jstarry — APPROVED** [review]

File: runtime/src/system_instruction_processor.rs (hunk 1 above)
Comment (joncinque) [reply runtime/src/system_instruction_processor.rs:188]: Yeah it looks a bit silly, but it avoids unnecessary calls to `try_account_ref_mut()` later on.  If there's no possible issues with doing those, I'm happy to get rid of this

Comment (jstarry) [reply runtime/src/system_instruction_processor.rs:188]: yeah no possible issues with that, let's get rid of it

File: runtime/src/system_instruction_processor.rs [hunk 3]
```diff
@@ -1103,7 +1125,7 @@ mod tests {
             0,
             &MockInvokeContext::new(vec![]),
         )
-        .is_ok(),);
+        .is_err(),);
```
Comment (t-nelson) [inline runtime/src/system_instruction_processor.rs:1128]: Can we `assert_eq!(transfer(), Err(InstructionError::MissingRequiredSignature));` here instead?  Reordering of errors will break consensus and we should be able to catch it.

File: runtime/src/system_instruction_processor.rs (hunk 2 above)
Comment (joncinque) [reply runtime/src/system_instruction_processor.rs:689]: I literally spent about an hour making sure that it does need to be a signer :sweat_smile:  I was worried that it was possible to create an ephemeral account without signing for it.  The tests are just written a bit strangely, so I wanted to clarify them

**Comment by joncinque** [comment]
@jstarry @t-nelson Could I ask for another quick check?  The stake split instruction uses the `create_account` with 0 lamports assumption, which makes one test fail in particular at https://github.com/solana-labs/solana/blob/be957f25c9f564b3937f7c42856aac3c00d52d9f/runtime/tests/stake.rs#L399.  In the test, we're splitting from a stake that is auto-authorized, so when we try to split from it, we get an error because an account with data is trying to transfer lamports.  I've updated the split instruction to allocate + assign separately.  Since this could break people using `split` with auto-authorized stakes in 1.6, I'll remove that backport and only backport to 1.7.  Let me know what you think.

File: runtime/src/system_instruction_processor.rs [hunk 4]
```diff
@@ -179,6 +179,7 @@ fn transfer_verified(
         ic_msg!(invoke_context, "Transfer: `from` must not carry data");
         return Err(InstructionError::InvalidArgument);
     }
+
```
Comment (mvines) [inline runtime/src/system_instruction_processor.rs:182]: nit: avoid change here

**Review by mvines — APPROVED** [review]

File: programs/stake/src/stake_instruction.rs [hunk 5]
```diff
@@ -261,13 +261,8 @@ pub fn split(
...
     vec![
-        system_instruction::create_account(
-            authorized_pubkey, // Sending 0, so any signer will suffice
-            split_stake_pubkey,
-            0,
-            std::mem::size_of::<StakeState>() as u64,
-            &id(),
-        ),
+        system_instruction::allocate(split_stake_pubkey, std::mem::size_of::<StakeState>() as u64),
+        system_instruction::assign(split_stake_pubkey, &id()),
```
Comment (t-nelson) [inline programs/stake/src/stake_instruction.rs:265]: This will break the Ledger app's parser.  Splits will need to be blind signed until it's updated

**Review by t-nelson — APPROVED** [review]

File: programs/stake/src/stake_instruction.rs (hunk 5 above)
Comment (joncinque) [reply programs/stake/src/stake_instruction.rs:265]: Looks like it used to be allocate + assign, then something changed https://github.com/LedgerHQ/app-solana/blob/1c72216edf4e5358f719b164a8d1b6100988b34d/libsol/transaction_printers.c#L29

Comment (t-nelson) [reply programs/stake/src/stake_instruction.rs:265]: Ah right.  Good ole lazy past-Trent saves the day :wink:

File: programs/stake/src/stake_instruction.rs [hunk 6]
```diff
@@ -288,7 +288,6 @@ pub fn split_with_seed(
             std::mem::size_of::<StakeState>() as u64,
             &id(),
         ),
-        system_instruction::assign_with_seed(split_stake_pubkey, base, seed, &id()),
```
Comment (t-nelson) [inline programs/stake/src/stake_instruction.rs]: Landmines everywhere!

https://github.com/solana-labs/solana/blob/7bb5f24274d1412e6f78616962d020a3de7aa841/runtime/src/system_instruction_processor.rs#L391-L404

Comment (joncinque) [reply programs/stake/src/stake_instruction.rs]: Seriously. You should see my most recent commit :sweat_smile:

File: programs/stake/src/stake_instruction.rs (hunk 5 above)
Comment (joncinque) [reply programs/stake/src/stake_instruction.rs:265]: That's a rarity.  I don't think past me has ever helped me that much

**Comment by joncinque** [comment]
@t-nelson @jstarry sorry, just one more check please on bdd6da6 -- with these changes, it's also possible to split a stake into an account that already has lamports, which is currently supported by the stake program.  For some reason, there's an existing test to make sure that this isn't possible, even though this behavior is reasonable.  As such, I've updated the test from a failure to a success.

**Comment by jstarry** [comment]
It's usually a mistake to split into an account which already has lamports. However, this is a recoverable mistake because a split account can have its extra lamports withdrawn and it can also be merged and closed completely if needed. I think that the "existing account" check should be a client side check in the cli rather than in the transaction. This looks good to me

**Comment by t-nelson** [comment]
> It's usually a mistake to split into an account which already has lamports. However, this is a recoverable mistake because a split account can have its extra lamports withdrawn and it can also be merged and closed completely if needed. I think that the "existing account" check should be a client side check in the cli rather than in the transaction. This looks good to me

Agreed. This is no worse than doing a system transfer into a delegated stake account. Which is annoying, but no funds are at risk past a bit of dust for fees

**Pull request merged**